### ISSUE TITLE
Instance doesn't behave like it used to for adaptable objects

### ISSUE THREAD

**Issue by rahulporuri** [enthought/traits#1211]
The `Instance` trait behavior has changed between traits 5.2.0 and traits 6.0.0:

Steps to reproduce : 
- Run this script in two environments, one with traits 5.2.0-1 and the other with traits 6.0.0-1
-- The script works fine with traits 5.2.0 but fails with 6.0.0-1

<details>

```
from traits.api import HasTraits, Instance, Interface, List, Str
from traits.adaptation.api import AdaptationOffer, get_global_adaptation_manager


class IData(Interface):
    name = Str


class DataWrap(HasTraits):
    item = Instance(IData)


class CustomData(HasTraits):
    name = Str


def adapt_custom_data_to_i_data(adaptee):
    return adapted


def main():
    am = get_global_adaptation_manager()
    am.register_offer(
        AdaptationOffer(
            factory=adapt_custom_data_to_i_data,
            from_protocol=CustomData,
            to_protocol=IData,
        )
    )

    cd = CustomData(name="data item 1")
    dl = DataWrap(item=cd)


if __name__ == "__main__":
    main()
```

</details>

The error observed with traits 6.0.0-1 is

```
> edm run -e traits-6.0.0 -- python traits-instance-mess.py
Traceback (most recent call last):
  File "traits-instance-mess.py", line 36, in <module>
    main()
  File "traits-instance-mess.py", line 32, in main
    dl = DataWrap(item=cd)
  File "C:\Users\rporuri\.edm\envs\traits-6.1.0\lib\site-packages\traits\base_trait_handler.py", line 75, in error
    object, name, self.full_info(object, name, value), value
traits.trait_errors.TraitError: The 'item' trait of a DataWrap instance must be an IData or None, but a value of <__main__.CustomData object at 0x000001D3F3C49C50> <class '__main__.CustomData'> was specified.
```

**Comment by rahulporuri** [comment]
If I used git bisect correctly, this change is caused by https://github.com/enthought/traits/pull/630 .

**Comment by mdickinson** [comment]
Yes, this was a deliberate behaviour change, as discussed in #630 and described in the release notes for Traits 6.0 (see the porting guide section of the 6.0.0rc0 notes). The solution is to declare that `CustomData` implements the `IData` interface using the `@provides` decorator.

**Comment by mdickinson** [comment]
@rahulporuri I'm a bit confused, though. The adaptation seems irrelevant here. Were you expecting `Instance` to do automatic adaptation? Because that doesn't happen in either Traits 5.2 _or_ Traits 6.0: your `adapt_custom_data_to_i_data` wouldn't get called with either version of Traits. (And if it did get called, it would give a `NameError` since `adapted` isn't defined.)

**Comment by mdickinson** [comment]
@rahulporuri I'm marking this as "not a bug", since the behaviour is intentional, but not closing until you've had a chance to respond to the comments. Is there something that you think Traits should be doing differently here?

**Comment by rahulporuri** [comment]
>  Were you expecting `Instance` to do automatic adaptation?

We weren't expecting `Instance` to do automatic adaptation - we were expecting it to infer that `CustomData` implicitly provided `IData` interface.

> (And if it did get called, it would give a NameError since adapted isn't defined.)

Sorry for that. You're right. The `AdaptationOffer` is irrelevant here.

**Comment by mdickinson** [comment]
> We weren't expecting `Instance` to do automatic adaptation - we were expecting it to infer that `CustomData` implicitly provided `IData` interface.

Okay. I'm really not keen on reversing this behaviour change; not only is it already released in Traits 6.0 and 6.1, but I believe that the new behaviour is a significant improvement on the old.

Previously, the *type's* interface was checked (via iterating method by method and comparing signatures) on every single assignment to something of the form `Instance(ISomeInterface)`. That's unexpected, inefficient, and the horribly dynamic interface checks don't always work as expected. It's much clearer if `Instance(ISomeInterface)` means that the assigned object must be an instance of `ISomeInterface` (i.e., passes an `isinstance(my_obj, ISomeInterface)` check). That's the obvious meaning, and is efficient and easy to implement with no peculiar corner cases.

For adaptation to `IDataItem`, there's a right way to do this, which is to explicitly mark the type of the adapted object as implementing `IDataItem`. For types that already implement `IDataItem` implicitly, there would be nothing wrong with registering them after the fact as such.

If we really need a trait type that does this sort of method-by-method signature comparison at every assignment, I'd much rather add a new trait type and keep `Instance` simple and obvious. But I doubt that we do need such a trait type.

**Comment by rahulporuri** [comment]
makes sense. please go ahead and close this issue
